conversion/conversion.go

```diff
@@ -0,0 +1,789 @@
+// Otherwise try to deduce the instance using gcloud.
+func GetInstance(project string, out *os.File) (string, error) {
+	l, err := getInstances(project)
+	if err != nil {
+		return "", err
+	}
+	if len(l) == 0 {
+		fmt.Fprintf(out, "Could not find any Spanner instances for project %s\n", project)
+		return "", fmt.Errorf("no Spanner instances for %s", project)
+	}
+	// Note: we could ask for user input to select/confirm which Spanner
+	// instance to use, but that interacts poorly with piping pg_dump/mysqldump data
+	// to the tool via stdin.
+	if len(l) == 1 {
+		fmt.Fprintf(out, "Using only available Spanner instance: %s\n", l[0])
+		return l[0], nil
+	}
+	fmt.Fprintf(out, "Available Spanner instances:\n")
+	for i, x := range l {
+		fmt.Fprintf(out, " %d) %s\n", i+1, x)
+	}
+	fmt.Fprintf(out, "Please pick one of the available instances and set the flag '--instance'\n\n")
+	return "", fmt.Errorf("auto-selection of instance failed: project %s has more than one Spanner instance. "+
+		"Please use the flag '--instance' to select an instance", project)
+}
+
+func getInstances(project string) ([]string, error) {
+	ctx := context.Background()
+	instanceClient, err := instance.NewInstanceAdminClient(ctx)
+	if err != nil {
+		return nil, analyzeError(err, project, "")
+	}
+	it := instanceClient.ListInstances(ctx, &instancepb.ListInstancesRequest{Parent: fmt.Sprintf("projects/%s", project)})
+	var l []string
+	for {
+		resp, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, analyzeError(err, project, "")
+		}
+		l = append(l, strings.TrimPrefix(resp.Name, fmt.Sprintf("projects/%s/instances/", project)))
+	}
+	return l, nil
+}
+
+func WriteSchemaFile(conv *internal.Conv, now time.Time, name string, out *os.File) {
+	f, err := os.Create(name)
+	if err != nil {
+		fmt.Fprintf(out, "Can't create schema file %s: %v\n", name, err)
+		return
+	}
+	// The schema file we write out includes comments, includes foreign keys
+	// and doesn't add backticks around table and column names. This file is
+	// intended for explanatory and documentation purposes, and is not strictly
+	// legal Cloud Spanner DDL (Cloud Spanner doesn't currently support comments).
+	// Change 'Comments' to false and 'ProtectIds' to true to write out a
+	// schema file that is legal Cloud Spanner DDL.
+	ddl := conv.GetDDL(ddl.Config{Comments: true, ProtectIds: false, ForeignKeys: true})
+	if len(ddl) == 0 {
+		ddl = []string{"\n-- Schema is empty -- no tables found\n"}
+	}
+	l := []string{
+		fmt.Sprintf("-- Schema generated %s\n", now.Format("2006-01-02 15:04:05")),
+		strings.Join(ddl, ";\n\n"),
+		"\n",
+	}
+	if _, err := f.WriteString(strings.Join(l, "")); err != nil {
+		fmt.Fprintf(out, "Can't write out schema file: %v\n", err)
+		return
+	}
+	fmt.Fprintf(out, "Wrote schema to file '%s'.\n", name)
+}
+
+func WriteSessionFile(conv *internal.Conv, now time.Time, name string, out *os.File) {
```

Comment:
All public functions should have a comment.

Reply:
added comment for public functions